### ISSUE TITLE
ERROR:host doesn't support passthrough of host PCI devices after vagrant_startup.sh

### ISSUE THREAD

**Issue by yusuke0324** [NVIDIA/deepops#291]
I'm trying to setup deepops on virtual machines on ubuntu 16.04.
I followed the tag 19.03 virtual instruction [here] (https://github.com/NVIDIA/deepops/tree/19.03/virtual) but got the following error after `$./vagrant_startup.sh`;
~~~
==> virtual-gpu01: Starting domain.
==> virtual-gpu01: An error occurred. The error will be shown after all tasks complete.
==> virtual-login: Machine already provisioned. Run `vagrant provision` or use the `--provision`
==> virtual-login: flag to force provisioning. Provisioners marked to run always will still run.
An error occurred while executing multiple actions in parallel.
Any errors that occurred are shown below.

An error occurred while executing the action on the 'virtual-gpu01'
machine. Please handle this error then try again:

There was an error talking to Libvirt. The error message is shown
below:

Call to virDomainCreateWithFlags failed: unsupported configuration: host doesn't support passthrough of host PCI devices
~~~

It used to be worked a couple of weeks ago but I noticed that all VMs are not running for some reason so I tried to resume them by `$ ./vagrant_shutdown.sh` and `./vagrant_startup.sh` but got the above error.

In my Vagrantfile, virtual-gpu01 looks like this;
~~~
  config.vm.define "virtual-gpu01" do |gpu|
    config.vm.provider "libvirt" do |v|
      v.memory = 16384
      v.cpus = 2
      v.machine_type = "q35"
      v.cpu_mode = "host-passthrough"
      # comment in for pci passthrough (and change bus according
      # to local hw setup - `lspci -nnk | grep NVIDIA`)
      v.pci :bus => '0x17', :slot => '0x00', :function => '0x1'
~~~
I tried comment out v.cpu_mode and v.pci but it threw the same errors.
When the VMs were running, I also got `Unable to connect to the server: dial tcp 10.0.0.2:6443: connect: no route to host` error by `./cluster_up.sh` on Host OS.
I'm not sure this is related to this issue. 
Do you have any idea how to solve this? Thanks

**Comment by michael-balint** [comment]
Can you confirm that you have followed the setup for [enabling virtualization and gpu passthrough](https://github.com/NVIDIA/deepops/blob/master/virtual/README.md#enabling-virtualization-and-gpu-passthrough) in the virtual readme? It sounds like you might have a BIOS setting (or some other setup) which is not enabled.

**Comment by yusuke0324** [comment]
I haven't followed the instruction. 
In BIOS changes, I could find Intel(VMX) virtualization option and Intel VT for Directoed I/O, and they are all `enabled`.
Could you explain a bit more detail about `Enable BIOS support for large-BAR1 GPUs: 'MMIO above 4G' or 'Above 4G encoding', etc.` What am I supposed to do with this description?

For `Bootloader changes`, I just exactly followed this instruction but I got an error after ` $sudo update-grub`;

~~~
$ sudo update-grub
Generating grub configuration file ...
Found linux image: /boot/vmlinuz-4.4.0-148-generic
Found initrd image: /boot/initrd.img-4.4.0-148-generic
Found linux image: /boot/vmlinuz-4.4.0-145-generic
Found initrd image: /boot/initrd.img-4.4.0-145-generic
Found linux image: /boot/vmlinuz-4.4.0-143-generic
Found initrd image: /boot/initrd.img-4.4.0-143-generic
Found linux image: /boot/vmlinuz-4.4.0-142-generic
Found initrd image: /boot/initrd.img-4.4.0-142-generic
Found linux image: /boot/vmlinuz-4.4.0-141-generic
Found initrd image: /boot/initrd.img-4.4.0-141-generic
Found linux image: /boot/vmlinuz-4.4.0-139-generic
Found initrd image: /boot/initrd.img-4.4.0-139-generic
Found memtest86+ image: /memtest86+.elf
Found memtest86+ image: /memtest86+.bin
  /run/lvm/lvmetad.socket: connect failed: No such file or directory
  WARNING: Failed to connect to lvmetad. Falling back to internal scanning.
done
~~~

I am not familiar with these settings, could you help me?

**Comment by michael-balint** [comment]
What hardware are you running on?

---

> Could you explain a bit more detail about Enable BIOS support for large-BAR1 GPUs: 'MMIO above 4G' or 'Above 4G encoding', etc. What am I supposed to do with this description?

You need to enable MMIO. This setting varies according to BIOS. If you can get output with the following command, you should be fine...

```
$ grep -oE 'svm|vmx' /proc/cpuinfo | uniq
vmx
```

(vmx for Intel or svm for AMD processors)

---

> I am not familiar with these settings, could you help me?

When you run `lspci -nnk -d 10de:1db1` (except substitute `10de:1db1` with your PCI ID obtained from `lspci -nn | grep NVIDIA`, what do you get? You should ideally see that `vfio-pci` is the driver in use. Otherwise, you need to ensure that you are following the steps in the doc (and debugging accordingly) to ensure that no other drivers are obtaining access to the GPUs before `vfio`. This is critical to the passthrough.

**Comment by yusuke0324** [comment]
@michael-balint  Thank you for your response.
This is my BIOS info, (got from `$sudo dmidecode | less`)
~~~
BIOS Information
        Vendor: American Megatrends Inc.
        Version: 1.2
        Release Date: 08/23/2018
        Address: 0xF0000
        Runtime Size: 64 kB
        ROM Size: 16384 kB
~~~

I got `vmx`
~~~
$ grep -oE 'svm|vmx' /proc/cpuinfo | uniq
vmx
~~~

~~~
$lspci -nn | grep NVIDIA
17:00.0 VGA compatible controller [0300]: NVIDIA Corporation GP102 [GeForce GTX 1080 Ti] [10de:1b06] (rev a1)
17:00.1 Audio device [0403]: NVIDIA Corporation GP102 HDMI Audio Controller [10de:10ef] (rev a1)
65:00.0 VGA compatible controller [0300]: NVIDIA Corporation GP102 [GeForce GTX 1080 Ti] [10de:1b06] (rev a1)
65:00.1 Audio device [0403]: NVIDIA Corporation GP102 HDMI Audio Controller [10de:10ef] (rev a1)
~~~

~~~
$lspci -nnk -d 10de:1b06
17:00.0 VGA compatible controller [0300]: NVIDIA Corporation GP102 [GeForce GTX 1080 Ti] [10de:1b06] (rev a1)
	Subsystem: Device [196e:1211]
	Kernel driver in use: nvidia
	Kernel modules: nvidiafb, nouveau, nvidia_410_drm, nvidia_410
65:00.0 VGA compatible controller [0300]: NVIDIA Corporation GP102 [GeForce GTX 1080 Ti] [10de:1b06] (rev a1)
	Subsystem: Device [196e:1211]
	Kernel driver in use: nvidia
	Kernel modules: nvidiafb, nouveau, nvidia_410_drm, nvidia_410
~~~

**Comment by michael-balint** [comment]
```Kernel driver in use: nvidia```

This is at least part of the issue - you need `vfio-pci` to be showing up here instead of nvidia. To accomplish this, [make the bootloader changes](https://github.com/NVIDIA/deepops/blob/master/virtual/README.md#bootloader-changes) and [blacklist the GPU devices](https://github.com/NVIDIA/deepops/blob/master/virtual/README.md#blacklist-the-gpu-devices). There are tests and sanity checks in the aforementioned docs to see if you are on the right path.

**Comment by michael-balint** [comment]
Since GPU passthrough is all-or-nothing (your machine is either in passthrough, or not), in most cases I've found it easiest to remove all GPU drivers rather than blacklist, since it can be tough to know what to explicitly blacklist. Then, `vfio-pci` should surely be the driver in use.

**Comment by yusuke0324** [comment]
I would like to use GPU as usual during I'm trying to set up deep ops with VMs.
Is it possible to switch the driver back and forth?  or, I have two GPUs on this machine so is it possible to provide one gpu to VMs and the others to Host OS?

**Comment by michael-balint** [comment]
What is your use case? Do you absolutely need to use VMs? Or would something like single node K8s work? What are you ultimately trying to do?

> Is it possible to switch the driver back and forth?

Yes, but you'd have to revert your changes to the bootloader and blacklisting. And when you want passthrough, you'd need to follow the instructions again.

> I have two GPUs on this machine so is it possible to provide one gpu to VMs and the others to Host OS?

No, the blacklisting / acquisition of the vfio-pci driver affects both cards or none.

**Comment by yusuke0324** [comment]
I would like to see what deepops can do and if deepops meets our requirements for our sharing servers with GPUs by setting up deepops my small workstation first.
(Our requirements are very simple like resource managing, job queue, and job scheduling.)
 
I usually use jupyterhub on docker container for GPU jobs. Once I switch the driver to vfio, do you think I can use jupyterhub on docker on VMs?

**Comment by michael-balint** [comment]
OK, so you want to run Kubernetes and deploy something like Kubeflow to launch your notebook containers? If so, I'd recommend doing the following instead of using virtual...

1. Use your laptop (or some other machine with network access to your compute node) to pull the deepops repo and run the setup script...

```sh
git clone git@github.com:NVIDIA/deepops.git
cd deepops
git checkout tags/19.03
./scripts/setup.sh
```

2. Next, copy the config and modify the inventory file to include only your compute node's ip...

```sh
cp config.example config

# Edit inventory and add nodes to the "KUBERNETES" section
# Note: Etcd requires an odd number of servers
vi config/inventory

# (optional) Modify `config/group_vars/*.yml` to set configuration parameters
```

3. Deploy Kubernetes...

```sh
# NOTE: If SSH requires a password, add: `-k`
# NOTE: If sudo on remote machine requires a password, add: `-K`
# NOTE: If SSH user is different than current user, add: `-u ubuntu`
ansible-playbook -l k8s-cluster playbooks/k8s-cluster.yml
```

4. Verify that the Kubernetes cluster is running. 

```sh
kubectl get nodes
``` 

Optionally, test a GPU job to ensure that your Kubernetes setup can tap into GPUs. 

```sh
kubectl run gpu-test --rm -t -i --restart=Never --image=nvidia/cuda --limits=nvidia.com/gpu=1 -- nvidia-smi
```

5. Install Helm...

```sh
./scripts/install_helm.sh
```

6. Deploy load balancer and ingress (optional but recommended)

```sh
./scripts/k8s_deploy_loadbalancer.sh
./scripts/k8s_deploy_ingress.sh
```

7. Deploy a Ceph cluster running on Kubernetes for services that require persistent storage (such as Kubeflow):

```sh
./scripts/k8s_deploy_rook.sh
```

Poll the Ceph status by running:

```sh
./scripts/ceph_poll.sh
```

8. Kubeflow is a popular way for multiple users to run ML workloads. It exposes a Jupyter Notebook interface where users can request access to GPUs via the browser GUI. Deploy Kubeflow with a convenient script:

```sh
./scripts/k8s_deploy_kubeflow.sh
```

**Comment by michael-balint** [comment]
Closing this, please open a new issue if you have a problem running DeepOps scripts as mentioned above.

**Comment by yusuke0324** [comment]
Jupyter notebook is not our usecase for deepops. (I just wanted to try to set up deepops on VMs on my workstation that has two gpu on ubuntsu but I usually use jupyter notebook on that machine with docker so I was worried if I can keep using the jupyter notebook as usual after changing PCI pass through. )
Our use case for deepops might be 
- Job scheduling
- resource management
- job queue visualization
We have several GPU servers with several GPUs in same network. We would like to share the GPUs within a team,

Thank you for the above instruction. Do you think the commands are available for Mac? My laptop is Mac OS but I got `/etc/os-release: No such file or directory` after `./scripts/setup.sh`.
